### ISSUE TITLE
impove parsing nal

### ISSUE THREAD

**Issue by marakew** [xtne6f/tsmemseg#1]
i think this algo can be improve as move index i by 3 if nal == 0,0,1

https://github.com/xtne6f/tsmemseg/blob/d600cb42148e7bb66cfb3afa7d32d79afe8fca03/mp4fragmenter.cpp#L62

**Comment by xtne6f** [comment]
A possible fix is as follows, for example. I welcome your pull request!

```C++
size_t nalPos = 0;
for (size_t i = 0; i < payloadSize; ++i) {
    if (i + 1 >= payloadSize || (payload[i] == 0 && payload[i + 1] == 0 && payload[i + 2] == 1)) {
        if (nalPos != 0) {
            onNalProc(payload + nalPos, (i + 1 >= payloadSize ? payloadSize : i - (payload[i - 1] == 0)) - nalPos);
        }
        nalPos = i + 3;
        i += 2;
    }
}
```

(Since the number of "0,0,1" in the payload is relatively rare, performance improvements may not be significant.)

**Comment by marakew** [comment]
how about this

```
    for (size_t i = 0; i + 1 < payloadSize; ) {
        if (i + 2 == payloadSize || (payload[i] == 0 && payload[i + 1] == 0
&& payload[i + 2] == 1)) {
            if (nalPos != 0) {
                onNalProc(payload + nalPos, (i + 2 == payloadSize ?
payloadSize : i - (payload[i - 1] == 0)) - nalPos);
            }
            i += 3;
            nalPos = i;
        } else
        {
              ++i;
        }
    }
```

вс, 3 мар. 2024 г. в 06:46, xtne6f ***@***.***>:

**Comment by marakew** [comment]
i mean move index below
basic idea take from
https://chromium.googlesource.com/external/webrtc/+/refs/heads/main/common_video/h264/h264_common.cc#20

**Comment by marakew** [comment]
```
    for (size_t i = 0; i < payloadSize; ) {
        if (i + 1 >= payloadSize || (payload[i] == 0 && payload[i + 1]
== 0 && payload[i + 2] == 1)) {
            if (nalPos != 0) {
                onNalProc(payload + nalPos, (i + 1 >= payloadSize ?
payloadSize : i - (payload[i - 1] == 0)) - nalPos);
            }
            i += 3;
            nalPos = i;
        } else
        if (payload[i + 2] > 0)
        {
           i += 3;
        } else
        {
              ++i;
        }
    }
```

**Comment by marakew** [comment]
```
    for (size_t i = 0; i < payloadSize; ) {
        if (i + 1 >= payloadSize || (payload[i] == 0 && payload[i + 1] == 0 && payload[i + 2] == 1)) {
            if (nalPos != 0) {
                onNalProc(payload + nalPos, (i + 1 >= payloadSize ? payloadSize : i - (payload[i - 1] == 0)) - nalPos);
            }
            i += 3;
            nalPos = i;
        } else
        if (payload[i + 2] > 0)
        {
           i += 3;
        } else
        {
              ++i;
        }
    }
```

**Comment by xtne6f** [comment]
Thank you, I had no idea that an approach like the BM method was available. I tested the performance of your proposed code.

The code for testing:
```c++
// Create pseudo NAL units payload
std::vector<uint8_t> payload;
std::random_device rng;
for (int i = 0; i < 8192; ++i) {
    payload.push_back(rng() & 0xff);
    // Start codes sometimes appear
    if (rng() % 100 == 0) {
        payload.push_back(0);
        payload.push_back(0);
        payload.push_back(1);
    }
}
payload.resize(8192);

// Measure performance
for (int i = 0; i < 3; ++i) {
    auto t = std::chrono::system_clock::now();
    for (int j = 0; j < 100000; ++j) {
        ParseNalsOld(payload.data(), payload.size(), [](const uint8_t *, size_t) {});
    }
    printf("ParseNalsOld = %lld msec\n", std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - t).count());
    t = std::chrono::system_clock::now();
    for (int j = 0; j < 100000; ++j) {
        ParseNalsImproved(payload.data(), payload.size(), [](const uint8_t *, size_t) {});
    }
    printf("ParseNalsImproved = %lld msec\n", std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - t).count());
}
```

And the result:
> ParseNalsOld = 985 msec
> ParseNalsImproved = 444 msec
> ParseNalsOld = 982 msec
> ParseNalsImproved = 450 msec
> ParseNalsOld = 987 msec
> ParseNalsImproved = 451 msec

Cool! There seems to be a performance improvement of about 2 times.
Could you send me a pull request?

**Comment by marakew** [comment]
any thanks to google webrtc 
the changes too small for PR
you able to make it from self

ps
look at place where key frame for HEVC
i have no link right now
but seems webrtc key frame detect as one of nal type
//BLA_W_LP = 16
//BLA_W_RADL = 17
//BLA_N_RADL = 18

//IDR_W_RADL = 19
//IDR_N_LP = 20
//CRA_NUT = 21

you are missed seems 16,17,18

**Comment by marakew** [comment]
about test
you seems missed about long nalu 0,0,0,1
and test only small 0,0,1

**Comment by marakew** [comment]
if you like tests
impove it
by convert to avc
add length before result nal unit
and compare result of old and new method
is there are result of nals equal

**Comment by xtne6f** [comment]
Thank you, I updated the test code. (a bit long, so I uploaded it to [gist]( https://gist.github.com/xtne6f/8e0fabbf94f042480ab7b92e233257f4 ). `if (i + 1 >= payloadSize ...` seem to be incorrect, so I fixed it (this is my mistake in the comment above).

Since I wasn't thinking about compiler optimizations, I examined the results again.

gcc -O0
> ParseNalsOld = 898 msec
> ParseNalsImproved = 573 msec
> ParseNalsOld = 909 msec
> ParseNalsImproved = 575 msec
> ParseNalsOld = 897 msec
> ParseNalsImproved = 568 msec

gcc -O1
> ParseNalsOld = 442 msec
> ParseNalsImproved = 284 msec
> ParseNalsOld = 445 msec
> ParseNalsImproved = 281 msec
> ParseNalsOld = 445 msec
> ParseNalsImproved = 275 msec

gcc -O2
> ParseNalsOld = 344 msec
> ParseNalsImproved = 742 msec
> ParseNalsOld = 347 msec
> ParseNalsImproved = 739 msec
> ParseNalsOld = 348 msec
> ParseNalsImproved = 752 msec

!?
This is an unexpected result. Hmm, the old code may be faster when optimized...

**Comment by marakew** [comment]
yes iam found where i known about hevc
https://github.com/video-dev/hls.js/blob/1599e8f1ff821665223034247432a4942493b852/src/demux/video/hevc-video-parser.ts#L68
but
avc keyframe not so simple too
they look into slice 
https://github.com/video-dev/hls.js/blob/1599e8f1ff821665223034247432a4942493b852/src/demux/video/avc-video-parser.ts#L51

hmm strange result of perfomance
jump predictor can't do something ?
i'll check with clang and cl

**Comment by marakew** [comment]
cl(msvc 2022 14.36.32532) release x64
```
ParseNalsOld = 4657 msec
ParseNalsImproved = 1970 msec
ParseNalsOld = 3927 msec
ParseNalsImproved = 2426 msec
ParseNalsOld = 4790 msec
ParseNalsImproved = 2582 msec
Entire count = 49200000
```
clang-cl(18.1.0) release x64
```
ParseNalsOld = 3308 msec
ParseNalsImproved = 1418 msec
ParseNalsOld = 3175 msec
ParseNalsImproved = 1803 msec
ParseNalsOld = 4870 msec
ParseNalsImproved = 2221 msec
Entire count = 51600000
```

**Comment by marakew** [comment]
yep
concat "Measure performance" with verification
and all will good
gcc optimize "Measure performance" with one "++n;" into one cache line

**Comment by marakew** [comment]
my test
https://godbolt.org/z/x5434v8ns

**Comment by marakew** [comment]
this produce seems little better perfomance

```
template<class P>
static void ParseNalsImproved(const uint8_t *payload, size_t payloadSize, P onNalProc)
{
    size_t nalPos = 0;
    for (size_t i = 0; ; ) {
        bool atEnd = i + 2 >= payloadSize;
        if (atEnd || (payload[i] == 0 && payload[i + 1] == 0 && payload[i + 2] == 1)) {
            if (nalPos != 0) {
                onNalProc(payload + nalPos, (atEnd ? payloadSize : i - (payload[i - 1] == 0)) - nalPos);
            }
            if (atEnd) {
                break;
            }
            i += 3;
            nalPos = i;
        } else
        if (payload[i + 2] > 0)
        {
            i += 3;
        } else
        {
            ++i;
        }
    }
}
```

**Comment by marakew** [comment]
iam increase round(4) and perfmormance round(600000)
add my implementation as 2
and reorder test to use my implementation cold cache
on windows
clang-cl x64
```
ParseNalsImproved2 = 23798 msec
ParseNalsImproved = 23494 msec
ParseNalsOld = 26210 msec
ParseNalsImproved2 = 21576 msec
ParseNalsImproved = 23568 msec
ParseNalsOld = 27391 msec
ParseNalsImproved2 = 21782 msec
ParseNalsImproved = 25771 msec
ParseNalsOld = 27822 msec
ParseNalsImproved2 = 27395 msec
ParseNalsImproved = 25857 msec
ParseNalsOld = 25727 msec
Entire count = 583200000
```
clang has strange optimization

msvc cl x64
```
ParseNalsImproved2 = 13341 msec
ParseNalsImproved = 17306 msec
ParseNalsOld = 30776 msec
ParseNalsImproved2 = 14933 msec
ParseNalsImproved = 18942 msec
ParseNalsOld = 30239 msec
ParseNalsImproved2 = 17691 msec
ParseNalsImproved = 19041 msec
ParseNalsOld = 29620 msec
ParseNalsImproved2 = 18114 msec
ParseNalsImproved = 19094 msec
ParseNalsOld = 28306 msec
Entire count = 568800000
```

so cl better then twice from clang-cl

i have no linux with last gcc for tests

**Comment by xtne6f** [comment]
I'm using GCC 13.2.0-x86-64, the result probably depends on the CPU. Interesting.
I also have an aarch64 (ARM) machine so I tested it (since very slow machine, reduced loop count to 10000) :

gcc 13.2.0 -O1 aarch64
> ParseNalsOld = 867 msec
> ParseNalsImproved = 425 msec
> ParseNalsOld = 867 msec
> ParseNalsImproved = 428 msec
> ParseNalsOld = 868 msec
> ParseNalsImproved = 425 msec

gcc 13.2.0 -O2 aarch64
> ParseNalsOld = 619 msec
> ParseNalsImproved = 408 msec
> ParseNalsOld = 618 msec
> ParseNalsImproved = 408 msec
> ParseNalsOld = 619 msec
> ParseNalsImproved = 408 msec

This is the expected result. Typically `ParseNalsImproved()` is likely to be faster, and GCC seems to generate slow code for some X64 machines due to cache hits etc.
I prefer `ParseNalsImproved()` as it is faster in more cases.

**Comment by marakew** [comment]
iam look inside the asm of clang and cl

cl create two indexs, from i
one i1 from i1=0 for buffer[i1] ....
second i2 from i2=2 for payloadSize

clang uses one index i

thats may be increase perfomance

**Comment by marakew** [comment]
may be will interesting
msvc cl pseudocode with two indexs

```
    do
    {
      nPos_v18 = 0i64;
      i0_v19 = 0i64;
      i2_v20 = 2i64;
      v21 = 0i64;
      while ( 1 )
      {
        while ( i2_v20 < payloadSize_v15
             && (payload_v14[i0_v19] || payload_v14[i0_v19 + 1] || payload_v14[i0_v19 + 2] != 1) )
        {
          if ( payload_v14[i0_v19 + 2] )
          {
            i0_v19 += 3i64;
            i2_v20 += 3i64;
          }
          else
          {
            ++i0_v19;
            ++i2_v20;
          }
          nal_v16 = vector_payload_v62[1];
        }

        if ( nPos_v18 )
        {
          if ( i2_v20 < payloadSize_v15 )
          {
            v22 = i0_v19 - (payload_v14[i0_v19 - 1] == 0);
            nal_v16 = vector_payload_v62[1];
          }
          else
          {
            v22 = payloadSize_v15;
          }

          nal_size_v23 = v22 - nPos_v18;
          nal_v24 = &payload_v14[nPos_v18];

          if ( v13
            || v21 >= (nal_v16 - v17) >> 3
            || (v25 = *(_QWORD *)(v17 + 8 * v21), nal_size_v23 != *(_QWORD *)(v66 + 8 * v21) - v25)
            || memcmp(nal_v24, &payload_v14[v25], nal_size_v23) )
          {
            v13 = 1;
          }
          ++v21;
          LODWORD(v72) = v72 + 1;
          nal_v16 = vector_payload_v62[1];
        }

        if ( i2_v20 >= payloadSize_v15 )
          break;

        i0_v19 += 3i64;
        nPos_v18 = i0_v19;
        i2_v20 += 3i64;
        nal_v16 = vector_payload_v62[1];
      }

      if ( v13 || v21 != (nal_v16 - v17) >> 3 )
        v13 = 1;
      v26 = v73-- == 1;
      nal_v16 = vector_payload_v62[1];
    }
    while ( !v26 );


```

**Comment by marakew** [comment]
clang pseudocode 
more optimized
but worse performance

```
    while ( 2 )
    {
      LODWORD(v83) = v25;
      v26 = 0i64;
      npos_v27 = 0i64;
      for ( i_ = 0i64; ; npos_v27 = i_ )
      {
        while ( 1 )
        {
          i_plus_2_v29 = i_ + 2;
          if ( i_ + 2 >= payloadSize_v8 )
          {
            nal_size_v30 = payloadSize_v8;
            if ( !npos_v27 )
              goto LABEL_24;
            goto LABEL_34;
          }
          if ( !payload_v4[i_] && !payload_v4[i_ + 1] && payload_v4[i_plus_2_v29] == 1 )
            break;
          i_ += 2i64 * (payload_v4[i_ + 2] != 0) + 1;
        }
        if ( !npos_v27 )
          goto SKIP_3_LABEL_39;
        nal_size_v30 = i_ - (payload_v4[i_ - 1] == 0);
LABEL_34:
        v31 = !v20;
        v20 = 1;

        if ( v31 && v26 < v23 )
        {
          v32 = nal_size_v30 - npos_v27;
          v33 = *(_QWORD *)(v22 + 8 * v26);
          if ( *(_QWORD *)(v24 + 8 * v26) - v33 == v32 )
          {
            v34 = v23;
            v35 = memcmp(&payload_v4[npos_v27], &payload_v4[v33], v32);
            v24 = v82;
            v23 = v34;
            v22 = v81;
            v20 = v35 != 0;
          }
        }
        ++v26;
        ++v19;
        if ( i_plus_2_v29 >= payloadSize_v8 )
          break;
SKIP_3_LABEL_39:
        i_ += 3i64;
      }
LABEL_24:
      v20 |= v26 != v23;
      v25 = v83 + 1;
      if ( (_DWORD)v83 != 599999 )
        continue;
      break;
    }

```

**Comment by marakew** [comment]
i think comparison (payload[i] == 0 && payload[i + 1] == 0 && payload[i + 2] == 1)
can be imroved with cast buffer[i] to uint16_t and compare with 0

there are too match comparison for branch prediction

**Comment by marakew** [comment]
changes for

`if (atEnd || (((payload[i]<<16) |(payload[i + 1]<<8) | (payload[i + 2])) == 1))`

imrove performance for clang by twice

**Comment by marakew** [comment]
clang like this, cl like that

```
template<class P>
static void ParseNalsImproved2(const uint8_t *payload, size_t payloadSize, P onNalProc)
{
    size_t nalPos = 0;
    for (size_t i = 0; ; ) {
        bool atEnd = i + 2 >= payloadSize;
#ifdef __clang__
        if (atEnd || (((payload[i]<<16) |(payload[i + 1]<<8) | (payload[i + 2])) == 1)) {
#else
        if (atEnd || (payload[i] == 0 && payload[i + 1] == 0 && payload[i + 2] == 1)) {
#endif
            if (nalPos != 0) {
                onNalProc(payload + nalPos, (atEnd ? payloadSize : i - (payload[i - 1] == 0)) - nalPos);
            }
            if (atEnd) {
                break;
            }
            i += 3;
            nalPos = i;
        } else
        if (payload[i + 2] > 0)
        {
            i += 3;
        } else
        {
            ++i;
        }
    }
}
```

this is clang output

```
ParseNalsOld = 3591 msec
ParseNalsImproved = 2468 msec
ParseNalsImproved2 = 911 msec

ParseNalsOld = 2771 msec
ParseNalsImproved = 3884 msec
ParseNalsImproved2 = 1593 msec

ParseNalsOld = 4088 msec
ParseNalsImproved = 3848 msec
ParseNalsImproved2 = 2242 msec

Entire count = 80100000
```

check what be better for gcc and imrove macros ifdef

**Comment by xtne6f** [comment]
The idea `if (atEnd || (((payload[i]<<16) |(payload[i + 1]<<8) | (payload[i + 2])) == 1))` (ParseNalsImproved2) was a little slower in my GCC environment...
Reading your pseudocode above, I thought it might be better to avoid always using `i` and `i+2` in the loop. That is, start the loop count from 2.

```C++
template<class P>
static void ParseNalsImproved3(const uint8_t *payload, size_t payloadSize, P onNalProc)
{
    size_t nalPos = 0;
    for (size_t i = 2; ; ) {
        if (i >= payloadSize || (payload[i] == 1 && payload[i - 1] == 0 && payload[i - 2] == 0)) {
            if (nalPos != 0) {
                onNalProc(payload + nalPos, (i >= payloadSize ? payloadSize : i - 2 - (payload[i - 3] == 0)) - nalPos);
            }
            if (i >= payloadSize) {
                break;
            }
            nalPos = i + 1;
            i += 3;
        } else
        if (payload[i] > 0)
        {
            i += 3;
        } else
        {
            ++i;
        }
    }
}
```

gcc -O1 x86-64
> ParseNalsOld = 559 msec
> ParseNalsImproved = 259 msec
> ParseNalsImproved2 = 798 msec
> ParseNalsImproved3 = 140 msec
> ParseNalsOld = 555 msec
> ParseNalsImproved = 254 msec
> ParseNalsImproved2 = 804 msec
> ParseNalsImproved3 = 145 msec
> ParseNalsOld = 569 msec
> ParseNalsImproved = 263 msec
> ParseNalsImproved2 = 816 msec
> ParseNalsImproved3 = 140 msec

gcc -O2 x86-64
> ParseNalsOld = 346 msec
> ParseNalsImproved = 736 msec
> ParseNalsImproved2 = 826 msec
> ParseNalsImproved3 = 315 msec
> ParseNalsOld = 349 msec
> ParseNalsImproved = 727 msec
> ParseNalsImproved2 = 823 msec
> ParseNalsImproved3 = 310 msec
> ParseNalsOld = 346 msec
> ParseNalsImproved = 727 msec
> ParseNalsImproved2 = 819 msec
> ParseNalsImproved3 = 313 msec

It's not bad. The unreasonable slowdown at O2 has been resolved.

**Comment by marakew** [comment]
ParseNalsImproved2 is my from above with macros for clang

some times cl doest like this improvements
but no so far
```
ParseNalsOld = 3227 msec
ParseNalsImproved = 1423 msec
ParseNalsImproved2 = 1420 msec
ParseNalsImproved3 = 1444 msec

ParseNalsOld = 4940 msec
ParseNalsImproved = 2611 msec
ParseNalsImproved2 = 2647 msec
ParseNalsImproved3 = 2707 msec

ParseNalsOld = 4388 msec
ParseNalsImproved = 2102 msec
ParseNalsImproved2 = 2158 msec
ParseNalsImproved3 = 2425 msec

Entire count = 76800000
```

clang likes
```
ParseNalsOld = 4788 msec
ParseNalsImproved = 4313 msec
ParseNalsImproved2 = 1889 msec
ParseNalsImproved3 = 1776 msec

ParseNalsOld = 4592 msec
ParseNalsImproved = 3817 msec
ParseNalsImproved2 = 2068 msec
ParseNalsImproved3 = 1509 msec

ParseNalsOld = 4407 msec
ParseNalsImproved = 3815 msec
ParseNalsImproved2 = 1947 msec
ParseNalsImproved3 = 2070 msec

Entire count = 98400000
```

as iam understand 
cl like two indexs

**Comment by xtne6f** [comment]
`ParseNalsImproved3` is not always the best performance, but it's better in many conditions. Plus, it's relatively simple. I finally decided to use it. (dcc47db14923186ad82478e5c3f633883d906060)
I also fixed the AVC keyframe determination. (0c661e192f89d3c76a6d8db06d682453175b7949)

**Comment by marakew** [comment]
great!
iam analyze few hecv streams
i see BLA always present at any PES
its intersting to get such stream hevc and avc when such strange keyframe use
for research

**Comment by marakew** [comment]
pls dont do git amend
this produce such git pull conflict

From https://github.com/xtne6f/tsmemseg
 + eca62f8...0c661e1 master     -> origin/master  (forced update)
Auto-merging mp4fragmenter.cpp
CONFLICT (content): Merge conflict in mp4fragmenter.cpp
Automatic merge failed; fix conflicts and then commit the result.

**Comment by xtne6f** [comment]
I made a careless mistake... Fixed. (5dc8323fb7d145ce8e7cbc7e735b052aacb62970)

I think BLA slices occur at edit points in open-GOP videos, it's strange that it appears so often. When BLA is present, the decoder (video player) should exclude associated RASL slices from presentation, so such BLA flags may affect playback results.

> pls dont do git amend

Oh! sorry about that. I have a habit of combining trivial changes into a single commit.
